osu.Game/Screens/SelectV2/SongSelect.cs

```diff
@@ -416,28 +435,54 @@ public void PresentScore(ScoreInfo score)
             this.Push(new SoloResultsScreen(score));
         }
 
+        /// <summary>
+        /// Finalises selection on the given <see cref="BeatmapInfo"/>.
+        /// </summary>
+        public void SelectAndStart(BeatmapInfo beatmap)
+        {
+            Beatmap.Value = beatmaps.GetWorkingBeatmap(beatmap);
+            OnStart();
+        }
+
         #region Beatmap management
 
-        public virtual bool EditingAllowed => false;
+        public IEnumerable<OsuMenuItem> GetForwardActions(BeatmapInfo beatmap)
+        {
+            foreach (var item in GetPrimaryActions(beatmap))
+                yield return item;
+
+            yield return new OsuMenuItemSpacer();
 
-        public void ManageCollections() => collectionsDialog?.Show();
+            foreach (var item in GetSecondaryActions(beatmap))
+                yield return item;
 
-        public void MarkPlayed(BeatmapInfo beatmap) => beatmaps.MarkPlayed(beatmap);
+            yield return new OsuMenuItemSpacer();
+        }
 
-        public void Hide(BeatmapInfo beatmap) => beatmaps.Hide(beatmap);
+        public virtual IEnumerable<OsuMenuItem> GetPrimaryActions(BeatmapInfo beatmap)
+        {
+            yield return new OsuMenuItem("Select", MenuItemType.Highlighted, () => SelectAndStart(beatmap))
+            {
+                Icon = FontAwesome.Solid.Check
+            };
+        }
```

Comment:
I like this even less. What is a "primary" action? What is a "secondary" action? Is the only distinction between them that they should have a menu item spacer between them?

This feels like overfitting abstractions. I'd take a plain declarative list of menu items, "duplicated" to the beatmap options popover, with a comment to make sure to keep the two places consistent any of the week over this.

This way of writing code is how horrible atrocities like `UserBasedTable` come to pass.

Reply:
I can work with copy paste if you're also okay with footer popover having screen specific checks in it. ie `if (SoloSongSelect) { include these options }`.

I will say that the code looks 100x worse to my eye than what I have here though.

Reply:
> What is a "primary" action? What is a "secondary" action?

Primary leads to another screen. Secondary does anything else.

Reply:
Can you explain why screen specific checks would be necessary?

If you want per-specific adjustments having `GetForwardActions()` virtual works for me. I'm not sure how that could not replace screen specific checks in footer popover.

Reply:
I'm just going on the fact that you're against that also existing above, so I assumed you're implying copy-paste in `PanelBeatmap` / `FooterPopover` / `PanelBeatmapStandalone` as it was before i refactored this.

Reply:
> Primary leads to another screen. Secondary does anything else.

I dunno if that functionally makes sense but if I'm to go with that then it should be in xmldoc.

Reply:
> I'm just going on the fact that you're against that also function existing

I can live with `GetForwardActions()` if you're confident it's not going to grow five weird carve-outs for specific song select subclasses further down the line. The "primary" / "secondary" thing I *really* would like gone, though. It's a nebulous abstraction, I won't buy those anymore from anyone.

Reply:
Can you check if [eec3ac2](https://github.com/ppy/osu/pull/33193/commits/eec3ac2feb979890fd25b7881b15c3be2e2e015c) looks better to your eye?

Reply:
I prefer that yes (aside from a potentially-hanging spacer at the end of list at https://github.com/ppy/osu/pull/33193/commits/eec3ac2feb979890fd25b7881b15c3be2e2e015c#diff-81b275c464303bf1452beae0752d55e076f92a818bc74ed4e900aceebfa57113R74).

Reply:
Great, I'll tidy things up and mark PR ready after a second pass.